### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by pedromlcosta**

### CODE REVIEW

File: fredit/lib/fredit_web/views/users/user_view.ex [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  def render("show.json", %{user: user}) do
+    %{data: render_one(user, UserView, "user.json")}
+  end
+
+  def render("user.json", %{user: user}) do
+    %{
+      id: user.id,
+      email: user.email,
+      username: user.username,
+      password_hash: user.password_hash,
```
Comment (speeddragon) [inline fredit/lib/fredit_web/views/users/user_view.ex]: Why you need to send password hash ?

File: fredit/priv/repo/migrations/20180821105411_create_users.exs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    create table(:users) do
+      add(:email, :string)
+      add(:username, :string)
+      add(:password_hash, :string)
+      add(:provider, :string)
+
+      timestamps()
+    end
+
+    create(index(:users, [:email], unique: true))
```
Comment (speeddragon) [inline fredit/priv/repo/migrations/20180821105411_create_users.exs]: Should username also be unique ?

File: fredit/priv/repo/seeds.exs [hunk 3]
```diff
@@ -9,3 +9,12 @@
 #
 # We recommend using the bang functions (`insert!`, `update!`
 # and so on) as they will fail if something goes wrong.
+
+alias Fredit.Repo
+
+Repo.insert!(%Fredit.Accounts.User{
+  email: "test@test.com",
+  password_hash: "$2y$12$6PEJYUh4vZlWl/vH8mGjzOa518y7a8AG/vHsDJ9aSuXnwvM5rmAzm",
```
Comment (speeddragon) [inline fredit/priv/repo/seeds.exs]: There isn't an issue with this, but only you know what is the password. You should provide the password (as a virtual field) and then convert it to an hash.

File: fredit/lib/fredit_web/controllers/info/info_controller.ex [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
+  plug(
+    FreditWeb.Plug.AuthPipeline
+    when action in [:info]
+  )
 
   def info(conn, _params) do
     {:ok, vsn} = :application.get_key(:fredit, :vsn)
     IO.puts(vsn)
-    IO.puts "+++"
+    IO.puts("+++")
```
Comment (speeddragon) [inline fredit/lib/fredit_web/controllers/info/info_controller.ex]: Remove it ?

File: fredit/lib/fredit_web/controllers/authentication/authentication_controller.ex [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      # used to create the access token
+      {:ok, %Fredit.Accounts.User{email: email, username: username, id: id} = user} ->
+        token_access =
+          Fredit.Guardian.Plug.sign_in(conn, user, %{}, token_type: "refresh")
+          |> Fredit.Guardian.Plug.current_token()
+
+        {:ok, _old_token, {token_refresh, _claims}} =
+          Fredit.Guardian.exchange(token_access, "refresh", "access")
+
+        IO.inspect(token_access)
```
Comment (speeddragon) [inline fredit/lib/fredit_web/controllers/authentication/authentication_controller.ex]: Remove this debug code.

File: fredit/lib/fredit/accounts/accounts.ex [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  def change_user(%User{} = user) do
+    User.changeset(user, %{})
+  end
+
+  def validate_login(email, password) do
+    with %User{} = user <- Repo.get_by(User, email: email),
+         true <- Comeonin.Bcrypt.checkpw(password, user.password_hash) do
+      {:ok, user}
+    else
+      _hi ->
```
Comment (speeddragon) [inline fredit/lib/fredit/accounts/accounts.ex]: Take a look into [`Bcrypt.dummy_checkpw/0`](https://hexdocs.pm/comeonin/Comeonin.Bcrypt.html#dummy_checkpw/1) when user doesn't exists.

File: fredit/lib/fredit_web/views/users/user_view.ex (hunk 1 above)
Comment (pedromlcosta) [reply fredit/lib/fredit_web/views/users/user_view.ex]: Fixing it. (Function was automatically created by Phoenix context, but it is not called anywhere)

**Pull request merged**